File: app/src/test/java/com/owncloud/android/utils/FilesSyncHelperTest.java
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2024 Benedek Major <benedek@major.onl>
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

package com.owncloud.android.utils;

import org.junit.Test;

import org.lukhnos.nnio.file.FileVisitResult;
import org.lukhnos.nnio.file.Path;
import org.lukhnos.nnio.file.SimpleFileVisitor;
import org.lukhnos.nnio.file.attribute.BasicFileAttributes;
import org.mockito.Mock;
import org.mockito.Mockito;


import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;

import static com.owncloud.android.utils.FilesSyncHelper.walkFileTreeRandomly;
import static org.junit.Assert.fail;
import static org.mockito.Mockito.mock;



public class FilesSyncHelperTest {
    @Mock
    private File mockFile;

    @Mock
    private Path mockPath;

    @Test
    public void testWalkFileTreeNullHandling(){
        mockFile = mock(File.class);
        mockPath = mock(Path.class);
        Mockito.when(mockFile.isDirectory()).thenReturn(true);
        Mockito.when(mockPath.toFile()).thenReturn(mockFile);
        Mockito.when(mockFile.listFiles()).thenReturn(null);
        Mockito.when(mockFile.canRead()).thenReturn(true);

        SimpleFileVisitor<Path> visitor = new SimpleFileVisitor<Path>() {
            @Override
            public FileVisitResult visitFile(Path path, BasicFileAttributes attrs) {
                return FileVisitResult.CONTINUE;
            }

            @Override
            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
                return FileVisitResult.CONTINUE;
            }

            @Override
            public FileVisitResult visitFileFailed(Path file, IOException exc) {
                return FileVisitResult.CONTINUE;
            }
        };
        try{
            walkFileTreeRandomly(mockPath,visitor);
        } catch (Exception e){
            StringWriter writer = new StringWriter();
            PrintWriter printWriter = new PrintWriter( writer );
            e.printStackTrace( printWriter );
            printWriter.flush();

            String stackTrace = writer.toString();
            fail("walkFileTree throws an exception: \n" + stackTrace);
        }
    }
    
}

File: app/src/test/java/com/owncloud/android/utils/FilesSyncHelperTest.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2026 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */
package com.owncloud.android.utils

import com.nextcloud.client.jobs.BackgroundJobManager
import com.owncloud.android.datamodel.SyncedFolder
import com.owncloud.android.datamodel.SyncedFolderProvider
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

class FilesSyncHelperTest {

    private val jobManager: BackgroundJobManager = mock()
    private val provider: SyncedFolderProvider = mock()
    private val enabled: SyncedFolder = mock()
    private val disabled: SyncedFolder = mock()

    @Before
    fun setup() {
        whenever(enabled.isEnabled) doReturn true
        whenever(disabled.isEnabled) doReturn false
        whenever(provider.syncedFolders) doReturn mutableListOf(enabled, disabled)
    }

    @Test
    fun `starts auto upload only for enabled folders, passing override power saving through`() {
        FilesSyncHelper.startAutoUploadForEnabledSyncedFolders(provider, jobManager, true)

        verify(jobManager).startAutoUpload(enabled, true)
        verify(jobManager, never()).startAutoUpload(eq(disabled), any())
    }
}
